src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/planning/SpaceBookingDayPlanner.kt

```diff
@@ -0,0 +1,147 @@
+   *   <li>Single occupancy bookings are never placed in rooms with other bookings</li>
+   *   <li>If there are multiple competing rooms (after above goals are satisfied), single occupancy bookings will
+   *   be placed into rooms with the least beds (ensuring minimum 'wastage' in these scenarios)</li>
+   * </ol>
+   *
+   * @param beds All beds available on the day being planned (i.e. won't include out of service beds or beds inactive according to start/end date)
+   * @param bookings All required bookings. The code assumes that the premise level characteristics have already been met
+   */
+  fun plan(
+    beds: Set<Bed>,
+    bookings: Set<SpaceBooking>,
+  ): DayPlannerResult {
+    val bedLedger = BedLedger(beds)
+    val planned = mutableListOf<DayPlannerPlannedBooking>()
+    val unplanned = mutableListOf<DayPlannerUnplannedBooking>()
+
+    val sortedBookings = bookings.toList()
+      .sortedByDescending { b -> b.requiredCharacteristics.sumOf { c -> c.weighting } }
+      .toMutableList()
+
+    sortedBookings.forEach { booking ->
+      val requiresSingleRoom = booking.requiredCharacteristics.any { it.singleRoom }
+      val characteristicsExcludingSingleRoom = booking.requiredCharacteristics.filter { !it.singleRoom }.toSet()
+
+      when (
+        val findResult = bedLedger.findBed(
+          characteristics = characteristicsExcludingSingleRoom,
+          requiresSingleRoom = requiresSingleRoom,
+        )
+      ) {
+        is FindBedResult.BedsFound -> {
+          findResult.beds.forEach { bed ->
+            planned.add(
+              DayPlannerPlannedBooking(
+                bed = bed,
+                booking = booking,
+              ),
+            )
+            bedLedger.reserve(bed)
+          }
+        }
+
+        is FindBedResult.BedNotFound -> {
+          unplanned.add(
+            DayPlannerUnplannedBooking(
+              booking = booking,
+            ),
+          )
+        }
+      }
+    }
+
+    return DayPlannerResult(
+      planned = planned.toList(),
+      unplanned = unplanned.toSet(),
+    )
+  }
+}
+
+private class BedLedger(initialState: Set<Bed>) {
+  val allBeds = initialState.sortedBy { it.label }.toMutableList()
+  val availableBeds = initialState.sortedBy { it.label }.toMutableList()
+  val reservedBeds = mutableListOf<Bed>()
+
+  fun findBed(
+    characteristics: Set<Characteristic>,
+    requiresSingleRoom: Boolean,
+  ): FindBedResult {
+    val bedsWithMatchingCharacteristics =
+      if (requiresSingleRoom) {
+        availableBeds
+          .filter { availableBed -> availableBed.room.characteristics.containsAll(characteristics) }
+          .filter { availableBed -> !isRoomOccupied(availableBed) }
+          .sortedBy { availableBeds -> allBedsInRoom(availableBeds.room).size }
+      } else {
+        availableBeds
+          .filter { availableBed -> availableBed.room.characteristics.containsAll(characteristics) }
+      }
```

Comment:
Could this logic be refactored into smaller functions, to make it simpler to read and test maybe?

e.g something like
 ```
private fun findMatchingBeds(
        characteristics: Set<Characteristic>,
        requiresSingleRoom: Boolean
    ): Set<Bed> {
        val matchingBeds = availableBeds.filter { bed ->
            bed.room.characteristics.containsAll(characteristics)
        }

        return if (requiresSingleRoom) {
            matchingBeds.filter { bed -> !isRoomOccupied(bed) }
                .sortedBy { bed -> bedsByRoom[bed.room]?.size ?: 0 }
        } else {
            matchingBeds
        }.toSet()
    }

    private fun findOptimalBed(candidates: Set<Bed>, requirements: Set<Characteristic>): Bed? =
        candidates.minByOrNull { bed ->
            bed.room.characteristics.minus(requirements).size
        }

val candidateBeds = findCandidateBeds(characteristics, requiresSingleRoom)
val optimalBed = findOptimalBed(candidateBeds, characteristics) ?: return FindBedResult.BedNotFound
        
       
FindBedResult.BedsFound(setOf(optimalBed))


```

Reply:
Apologies for the terrible formatting

Reply:
This is great ... i've mostly gone with code proposed as-is. I agree breaking it out into chunks makes it easier to understand. I've stuck with `bedWithLeastSurplusCharacteristics` instead of `optimalBed` as i think it makes it clearer what the optimisation is (it's not so obvious reading the code)